Add "index" command to Atomic App

cdrage opened the pull request:
This adds an index command to list multiple Nulecule's from an external
library that contains an index.yaml file in it's root directory.

Built upon @vpavlin 's work :)

TODO:
- [x] `atomicapp index list --update` command
- [x] "official" index docker image (will need to add this to projectatomic/nulecule-library)
- [x] script to generate index.yaml
- [x] integration tests
- [x] docs going to update via https://github.com/projectatomic/atomicapp/issues/768

cdrage commented:
Here's what it looks like right now if you pull the PR and run `atomicapp index list`:

```sh
dropbox/dev/atomicapp  add-index ✔                                                                                                                                                                                                                       2m  ⍉
▶ atomicapp index list
ID                 VER      PROVIDERS  LOCATION                                
redis-atomicapp    0.0.1    {D,K}      docker://projectatomic/redis-atomicapp  
mariadb-atomicapp  1.0.0    {D,O,K}    docker://projectatomic/mariadb-atomicapp
helloapache-app    0.0.1    {D,K}      docker://projectatomic/helloapache   
```

cdrage commented:
https://github.com/projectatomic/nulecule-library/pull/96 location metadata has been added for updating the information (including docker.io urls)

cdrage commented:
Hey guys.

Pretty much done at this point in terms of Atomic App code.

Here are the updates

- Not going to include `atomicapp index info` for now until I research more on how the output format should look like, also whether or not we should include `atomicapp index edit`
- Started work on tests, have two more to go
- Need to add docs not only on how to use the command, but also that there will now be an `.atomicapp/index` on your system
- Haven't tested with `atomic cli` functionality yet

In order to try out this PR, you must:
```
# with https://github.com/projectatomic/nulecule-library/pull/97 PR patched
cd nulecule-library
docker build -t projectatomic/nulecule-library .
# then
atomicapp index list # should auto-update (put the index.yaml into your .atomicapp/index.yaml
```

Hope you guys like it :)
@kadel @rtnpro @dustymabe @vpavlin

cdrage commented:
What I'm waiting for/block with is:

- https://github.com/projectatomic/nulecule-library/pull/97 to be merged
- hub.docker.com to be updated with projectatomic/nulecule-library docker image for people to pull
- code reviews

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,204 @@
+        # eliminating the INFO :: Atomic App: X.X.X - Mode: Index
+        # logging information. If we passed --debug, ignore.
+        # if not logging.getLogger().getEffectiveLevel() == logging.DEBUG:
+        #     sys.stdout.write("\033[F")
```
remove?

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,204 @@
+            anymarkup.serialize_file(self.index, output_location, format="yaml")
+        logger.info("index.yaml generated")
+
+    def _fetch(self, index_image=INDEX_IMAGE):
```
would prefer for this funcname to be more specific `fetch_index_container` or something along those lines.

dustymabe replied:
I don't see any place where _fetch() is called with an argument. should we not enable someone to specify a different image to use as an index if they prefer?

cdrage replied:
I'll add this as a CLI command. But yes, by default it'll pull the atomicapp one.

dustymabe replied:
did you add the option to specify a different container name to the CLI? I don't see it.

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,204 @@
+
+        dh = DockerHandler()
+        dh.pull(index_image)
+        dh.extract(index_image, "/index.yaml", self.index_location, file=True)
```
shouldn't we be using `INDEX_NAME` here?

dustymabe commented on atomicapp/constants.py:
```diff
@@ -82,3 +82,10 @@
+INDEX_LOCATION = ".atomicapp/index.yaml"
+INDEX_NAME = "index.yaml"
+INDEX_DEFAULT_IMAGE_LOCATION = "localhost"
+INDEX_GEN_DEFAULT_OUTPUT_LOC = "./index.yaml"
```
many of these variables essentially contain the same information. should we somehow have INDEX_NAME incorporated in the other ones that currently have `index.yaml` hardcoded?

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,204 @@
+
+        providers_set = set()
+        for component in nulecule.components:
+            if component.artifacts:
```
Should there ever be a case where no artifacts exist?

cdrage replied:
Hopefully not. But I'll add this is a race condition.

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,204 @@
+        for component in nulecule.components:
+            if component.artifacts:
+                if len(providers_set) == 0:
+                    providers_set = set(component.artifacts.keys())
```
does this properly handle the "inherit" case for kube/openshift?

cdrage replied:
Yes. For example (the redis example):
```
redis-atomicapp           0.0.1    {D,O,K}    docker.io/projectatomic/redis-centos7-atomicapp
```

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,204 @@
+            raise Exception("Location must be a git repository")
+
+        for f in os.listdir(location):
+            nulecule_dir = os.path.join(location, f)
```
so every dir in the `location` is a nulecule_dir? we just assume this?

cdrage replied:
Yes. We assume each directory has a Nulecule component for the generation command.

From your comment down below (in regards to the Git dependency) we can instead create a generation script inside of the projectatomic/nulecule-library repo instead and further improve it.

dustymabe replied:
so the problem with assuming it is a nulecule dir is: will you get an exception if it's not ? we need to fail gracefully. can you test and make sure?

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,204 @@
+            if f.startswith("."):
+                continue
+            if os.path.isdir(nulecule_dir):
+                index_info = self._nulecule_get_info(nulecule_dir)
```
what if nulecule_dir doesn't actually have a Nulecule in it?

cdrage replied:
Good question :)

cdrage replied:
This actually fails already:
```
ERROR  :: No Nulecule file exists in directory: ./gocounter-scratch-atomicapp
```

:)

dustymabe replied:
ahh ok - so does that mean if there is one directory in the whole repo that doesn't have a nulecule then the whole process will fail?

cdrage replied:
@dustymabe yes. Should we "warn" instead?

dustymabe replied:
yeah I would just warn (or maybe even INFO) and say no nulecule file within dir.. skipping

dustymabe commented on atomicapp/nulecule/container.py:
```diff
@@ -155,9 +155,12 @@ def extract(self, image, source, dest, update=False):
+        if file:
+            Utils.copy_file(src, dest)
+        else:
+            Utils.copy_dir(src, dest, update)
```
the two statements below need to be unindented. the tmpdir needs to be deleted regardless of if we were doing just a file copy or an entire dir copy.

dustymabe commented on atomicapp/nulecule/container.py:
```diff
@@ -90,7 +90,7 @@ def pull(self, image, update=False):
         cockpit_logger.info('Skipping pulling docker image: %s' % image)
 
-    def extract(self, image, source, dest, update=False):
+    def extract(self, image, source, dest, update=False, file=False):
```
ok after some thought I think we need to refactor this function. unfortunately this isn't a generic function for copying out data from a container. It appears that way but we have all of the `App exists locally. Performing update...` stuff in the middle. This should be broken out into two functions. 

- One. This function just copies a file or directory out of a container to a target file or directory (the specified directory could be a tmpdir that will be later cleaned up by the calling function). We may even choose to put this function in a different file as it may belong somewhere else.
- Two. This function creates a tmp dir, calls the first function, does the `App exists locally. Performing update...` and then cleans up the tmp dir. 

Does that make sense?

cdrage replied:
That makes sense! I'll implement it.

cdrage replied:
Bit confused on the second function, since we don't detect explicit whether something needs to be updated, we just pass in True/False to the `update` variable.

So the second function is just going to call the first and override the previous information?

dustymabe replied:
can you fpaste or pastebin what you have right now (the two new functions) and I'll try to understand your confusion and then try to help come up with a solution

dustymabe replied:
@cdrage - here is a 5 minute try at this: http://paste.fedoraproject.org/389012/46799478

I would suggest we do this in a separate PR and then rebase this PR on top.

dustymabe commented on requirements.txt:
```diff
@@ -2,3 +2,4 @@ anymarkup
 jsonpointer
 requests
 websocket-client
+GitPython
```
I'm not a big fan of adding a new dependency for something that is probably going to get used 1% of the time. hmmm

dustymabe replied:
at least I think it is only used for the "generate" portion of the code right? so that means it probably won't get used that much compared to the other functionality

cdrage replied:
I agree @dustymabe in regards to the extra dependency. This call will be rarely used. Only used if the person wants to create their own "nulecule library" hosted on a different docker container, etc.

Do you agree that we should instead take the `generation` command and add it to `projectatomic/nulecule-library` repo?

dustymabe replied:
Another alternative is to just only support generation from a local directory and  have the user run `git clone` before doing a "generation".

cdrage replied:
Ah.... that too xD we can just say that we support paths only.

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
@dustymabe other than the TODO I added (going to work on adding a test + failure if there is no `Nulecule` in the directory when using generate, this is good to go.

dustymabe commented:
@cdrage - i'll try to look at this tomorrow - also you still have one check box for docs

cdrage commented:
Okay this has been update and ready for code review + possible merging.

@dustymabe @rtnpro @containscafeine @kadel if ya'll want to try it out / review it :)

cdrage commented:
I'll be updating the docs in a separate PR since we need to update https://github.com/projectatomic/atomicapp/blob/master/docs/cli.md anyways.
Opened up issue #768 for this.

cdrage commented:
#dotests

dustymabe commented on atomicapp/cli/main.py:
```diff
@@ -381,6 +395,28 @@ def create_parser(self):
+
+        index_list = index_action.add_parser("list")
+        index_list.add_argument(
+            "--update",
```
so what do you think about having this as a separate subcommand rather than an option to `--list`? seems like `atomicapp index update` would be more natural than `atomicapp index list --update`.

vpavlin replied:
How about `atomicapp index pull` ? (It's a docker pull anyway:) )

cdrage replied:
That'd be good, what do you think @dustymabe ?

dustymabe replied:
ehh i like "update" but don't really care.. potato potato - go with whatever

dustymabe commented on atomicapp/utils.py:
```diff
@@ -372,6 +373,11 @@ def copy_dir(src, dest, update=False, dryrun=False):
             distutils.dir_util.copy_tree(src, dest, update)
 
     @staticmethod
+    def copy_file(src, dest, dryrun=False):
```
did you forget to add a file to the commit? I don't see where copy_file gets used anywhere.

cdrage replied:
You're right. Updated this.

cdrage commented:
@dustymabe 

updated based on your comments. added the `atomicapp index update` command (i liked your suggestion better). removed the random `utils.py` stuff I still had there as well.

cdrage commented:
#dotests

dustymabe commented on atomicapp/cli/main.py:
```diff
@@ -381,6 +394,25 @@ def create_parser(self):
+        index_generate.add_argument(
+            "location",
+            help=(
+                "Path or Git repository link containing Nulecule applications "
```
'or Git repository' - is that still true?

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,203 @@
+                try:
+                    index_info = self._nulecule_get_info(nulecule_dir)
+                except NuleculeException as e:
+                    logger.warning("SKIPPING %s. %s" %
```
just a little more context would be nice.. "No Nulecule found in /path/to/dir, Skipping.." 

Doesn't have to be that, just a suggestion.

cdrage replied:
I left it out since the error out from `base.py` does that already: 
```
raise NuleculeException("No Nulecule file exists in directory: %s" % src)
```

dustymabe replied:
ok

dustymabe commented on atomicapp/index.py:
```diff
@@ -0,0 +1,203 @@
+
+        dh = DockerHandler()
+        dh.pull(index_image)
+        dh.extract_files(index_image, "/" + INDEX_NAME, self.index_location)
```
hmm ok - so running this in the "in container" use case doesn't work. It doesn't copy files to ~/.atomicapp/ - it copies them to that dir in the container but not outside the container.

one thing to consider is do we need to store things on the machine or could we just leave the index in the container and cp it out each time the user runs the command.

cdrage replied:
@dustymabe 
Could we use the `utils.py` function `getUserHome()` to set the file location appropriately?

dustymabe replied:
yes. i think that is what that function is for, but we should also make sure the file isn't owned by root after we get done creating it. right?

cdrage commented:
Hey @dustymabe

So the index_location has been updated in the `init` of the `index.py` file to the proper location via:
```
self.index_location = os.path.join(Utils.getUserHome(), INDEX_LOCATION)
```

```
Utils.setFileOwnerGroup(dest)
```
has also been added to the `extract_files` function so it properly sets the permission when pulling.

One last review? :)

cdrage commented:
```
dropbox/dev/atomicapp  add-index ✔                                                                                                                                                                                                                        1m  
▶ atomicapp index list
INFO   :: Atomic App: 0.6.1 - Mode: Index
ID                        VER      PROVIDERS  LOCATION                                             
postgresql-atomicapp      1.0.0    {D,O,K}    docker.io/projectatomic/postgresql-centos7-atomicapp 
flask_redis_nulecule      0.0.1    {D,K}      docker.io/projectatomic/flask-redis-centos7-atomicapp
redis-atomicapp           0.0.1    {D,O,K}    docker.io/projectatomic/redis-centos7-atomicapp      
gocounter                 0.0.1    {D,K}      docker.io/projectatomic/gocounter-scratch-atomicapp  
mariadb-atomicapp         1.0.0    {D,O,K}    docker.io/projectatomic/mariadb-centos7-atomicapp    
helloapache-app           0.0.1    {D,K,M}    docker.io/projectatomic/helloapache                  
mongodb-atomicapp         1.0.0    {D,O,K}    docker.io/projectatomic/mongodb-centos7-atomicapp    
etherpad-app              0.0.1    {D,O,K}    docker.io/projectatomic/etherpad-centos7-atomicapp   
apache-centos7-atomicapp  0.0.1    {D,K,M}    docker.io/projectatomic/apache-centos7-atomicapp     
wordpress-atomicapp       2.0.0    {D,O,K}    docker.io/projectatomic/wordpress-centos7-atomicapp  
skydns-atomicapp          0.0.1    {K}        docker.io/projectatomic/skydns-atomicapp             
guestbookgo-atomicapp     0.0.1    {O,K}      docker.io/projectatomic/guestbookgo-atomicapp        
mariadb-app               0.0.1    {D,K}      docker.io/projectatomic/mariadb-fedora-atomicapp     
gitlab-atomicapp          1.2.0    {D,K}      docker.io/projectatomic/gitlab-centos7-atomicapp     

dropbox/dev/atomicapp  add-index ✔                                                                                                                                                                                                                        1m  
▶ atomicapp index update
INFO   :: Atomic App: 0.6.1 - Mode: Index
INFO   :: Updating the index list
INFO   :: Pulling latest index image...
INFO   :: Skipping pulling docker image: projectatomic/nulecule-library
INFO   :: Copying files from image projectatomic/nulecule-library:/index.yaml to /home/wikus/.atomicapp/index.yaml
INFO   :: Index updated
```

And the permissions:
```
▶ tree ~/.atomicapp -ug 
/home/wikus/.atomicapp
└── [wikus    wikus   ]  index.yaml

0 directories, 1 file
```

dustymabe commented:
LGTM

cdrage commented:
#dotests
